File: backends/ktor/src/main/kotlin/com/hellotimepro/ktor/service/LlmClient.kt
package com.hellotimepro.ktor.service

import com.hellotimepro.ktor.config.AppConfig
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import kotlinx.serialization.json.putJsonObject
import org.slf4j.LoggerFactory
import java.io.IOException
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

/**
 * 结构化 JSON 生成的 LLM 客户端。日志规范、网关重试、CF 1010 等坑见 docs/dev-notes.md §3。
 * 契约测试默认 LLM_ENABLED=false，会在此立即抛异常，由上层走本地兜底；HTTP 路径不在测试链上。
 */
class LlmClient(private val config: AppConfig) {
    private val log = LoggerFactory.getLogger(LlmClient::class.java)
    private val json = Json { ignoreUnknownKeys = true }

    class LlmClientException(message: String, val status: Int = 0, cause: Throwable? = null) :
        RuntimeException(message, cause)

    data class SchemaSpec(
        val schemaName: String,
        val schema: JsonObject,
        val systemPrompt: String,
        val maxOutputTokens: Int,
        val maxTokens: Int,
    )

    fun generateCapsuleSuggestion(prompt: String): JsonObject =
        generateStructuredJson(prompt, SUGGESTION_SPEC)

    fun generateCapsuleRecommendations(prompt: String): JsonObject =
        generateStructuredJson(prompt, RECOMMENDATION_SPEC)

    private fun generateStructuredJson(prompt: String, spec: SchemaSpec): JsonObject {
        val llm = config.llm
        if (!llm.enabled || llm.apiKey.isBlank()) {
            throw LlmClientException("LLM is disabled or missing API key")
        }
        return when (llm.apiStyle) {
            "responses" -> generateWithResponses(prompt, spec)
            "auto" -> try {
                generateWithResponses(prompt, spec)
            } catch (e: LlmClientException) {
                log.info("Responses API unavailable ({}); falling back to chat completions", e.message)
                generateWithChat(prompt, spec, disableThinking = true)
            }
            else -> generateWithChat(prompt, spec, disableThinking = true)
        }
    }

    private fun generateWithResponses(prompt: String, spec: SchemaSpec): JsonObject {
        val payload = buildJsonObject {
            put("model", config.llm.model)
            put("input", prompt)
            put("max_output_tokens", spec.maxOutputTokens)
            putJsonObject("text") {
                putJsonObject("format") {
                    put("type", "json_schema")
                    put("name", spec.schemaName)
                    put("strict", true)
                    put("schema", spec.schema)
                }
            }
        }
        return parseJsonObject(extractResponsesText(postJson(responsesUrl(), payload)))
    }

    private fun generateWithChat(prompt: String, spec: SchemaSpec, disableThinking: Boolean): JsonObject {
        val payload = buildJsonObject {
            put("model", config.llm.model)
            putJsonArray("messages") {
                add(buildJsonObject { put("role", "system"); put("content", spec.systemPrompt) })
                add(buildJsonObject { put("role", "user"); put("content", prompt) })
            }
            put("max_tokens", spec.maxTokens)
            if (disableThinking) putJsonObject("thinking") { put("type", "disabled") }
        }
        return try {
            parseJsonObject(extractChatText(postJson(chatUrl(), payload)))
        } catch (e: LlmClientException) {
            if (e.status != 400) throw e
            // 某些网关不认 thinking 字段，去掉重试一次。
            parseJsonObject(extractChatText(postJson(chatUrl(), buildJsonObject {
                put("model", config.llm.model)
                putJsonArray("messages") {
                    add(buildJsonObject { put("role", "system"); put("content", spec.systemPrompt) })
                    add(buildJsonObject { put("role", "user"); put("content", prompt) })
                }
                put("max_tokens", spec.maxTokens)
            })))
        }
    }

    /** 向 url POST JSON；瞬时网络/TLS 错误按配置重试，HTTP 4xx/5xx 与坏 JSON 不重试。 */
    private fun postJson(url: String, payload: JsonObject): JsonObject {
        val llm = config.llm
        val body = json.encodeToString(JsonObject.serializer(), payload)
        val client = HttpClient.newBuilder().connectTimeout(Duration.ofMillis(llm.timeoutMs)).build()
        val attempts = maxOf(1, llm.maxRetries + 1)
        var lastError: Exception? = null

        for (attempt in 1..attempts) {
            log.info("LLM request  model={} url={} attempt={}/{}", llm.model, url, attempt, attempts)
            val start = System.currentTimeMillis()
            try {
                val request = HttpRequest.newBuilder(URI.create(url))
                    .timeout(Duration.ofMillis(llm.timeoutMs))
                    .header("Authorization", "Bearer ${llm.apiKey}")
                    .header("Content-Type", "application/json")
                    .header("Accept", "application/json")
                    .header("User-Agent", llm.userAgent)
                    .POST(HttpRequest.BodyPublishers.ofString(body))
                    .build()
                val response = client.send(request, HttpResponse.BodyHandlers.ofString())
                val elapsed = System.currentTimeMillis() - start
                val status = response.statusCode()
                if (status < 200 || status >= 300) {
                    log.warn("LLM error    model={} elapsed_ms={} status={}", llm.model, elapsed, status)
                    throw LlmClientException("HTTP $status: ${response.body()}", status = status)
                }
                val parsed = try {
                    json.parseToJsonElement(response.body()).jsonObject
                } catch (e: Exception) {
                    log.warn("LLM error    model={} elapsed_ms={} error=invalid-json", llm.model, elapsed)
                    throw LlmClientException("LLM response was not valid JSON", cause = e)
                }
                log.info("LLM response model={} elapsed_ms={} tokens={}", llm.model, elapsed, extractTokens(parsed))
                return parsed
            } catch (e: IOException) {
                val elapsed = System.currentTimeMillis() - start
                val willRetry = attempt < attempts
                log.warn(
                    "LLM error    model={} elapsed_ms={} error={}{}",
                    llm.model, elapsed, e.message, if (willRetry) " (will retry)" else "",
                )
                lastError = e
                if (willRetry) Thread.sleep(llm.retryBackoffMs * attempt)
            } catch (e: InterruptedException) {
                Thread.currentThread().interrupt()
                throw LlmClientException("LLM request interrupted", cause = e)
            }
        }
        throw LlmClientException(lastError?.message ?: "LLM request failed", cause = lastError)
    }

    private fun extractTokens(body: JsonObject): String {
        val usage = body["usage"]?.jsonObject ?: return "n/a"
        (usage["total_tokens"]?.jsonPrimitive?.content?.toLongOrNull())?.let { if (it > 0) return it.toString() }
        val sum = (usage["input_tokens"]?.jsonPrimitive?.content?.toLongOrNull() ?: 0) +
            (usage["output_tokens"]?.jsonPrimitive?.content?.toLongOrNull() ?: 0)
        return if (sum > 0) sum.toString() else "n/a"
    }

    private fun extractChatText(body: JsonObject): String {
        val choices = body["choices"]?.jsonArray ?: throw LlmClientException("LLM chat response missing choices")
        val content = choices.firstOrNull()?.jsonObject
            ?.get("message")?.jsonObject?.get("content")?.jsonPrimitive?.contentOrNullSafe()
        if (content.isNullOrBlank()) throw LlmClientException("LLM chat response missing content")
        return content
    }

    private fun extractResponsesText(body: JsonObject): String {
        (body["output_text"] as? JsonPrimitive)?.contentOrNullSafe()?.let { if (it.isNotBlank()) return it }
        body["output"]?.jsonArray?.forEach { item ->
            item.jsonObject["content"]?.jsonArray?.forEach { entry ->
                (entry.jsonObject["text"] as? JsonPrimitive)?.contentOrNullSafe()?.let {
                    if (it.isNotBlank()) return it
                }
            }
        }
        throw LlmClientException("LLM response did not contain output text")
    }

    private fun parseJsonObject(raw: String): JsonObject {
        var text = raw.trim()
        if (text.startsWith("```")) {
            text = text.replace(Regex("^```[a-zA-Z]*\\s*"), "").replace(Regex("\\s*```$"), "").trim()
        }
        return try {
            json.parseToJsonElement(text).jsonObject
        } catch (_: Exception) {
            val start = text.indexOf('{')
            val end = text.lastIndexOf('}')
            if (start < 0 || end <= start) throw LlmClientException("LLM output was not valid JSON")
            json.parseToJsonElement(text.substring(start, end + 1)).jsonObject
        }
    }

    private fun JsonPrimitive.contentOrNullSafe(): String? = if (this.isString) this.content else this.content

    private fun stripTrailingSlash(value: String): String =
        if (value.isBlank()) "https://api.openai.com/v1"
        else value.trimEnd('/')

    private fun responsesUrl(): String = stripTrailingSlash(config.llm.baseUrl) + "/responses"
    private fun chatUrl(): String = stripTrailingSlash(config.llm.baseUrl) + "/chat/completions"

    companion object {
        private val SUGGESTION_SPEC = LlmClient.SchemaSpec(
            schemaName = "capsule_suggestion",
            schema = buildJsonObject {
                put("type", "object")
                put("additionalProperties", false)
                putJsonArray("required") { add("title"); add("content"); add("openInDays") }
                putJsonObject("properties") {
                    putJsonObject("title") { put("type", "string") }
                    putJsonObject("content") { put("type", "string") }
                    putJsonObject("openInDays") { put("type", "integer"); put("minimum", 1); put("maximum", 3650) }
                }
            },
            systemPrompt = "你只返回严格 JSON 对象，不要 Markdown、代码块或解释。" +
                "JSON 必须包含字符串字段 title、content 和整数字段 openInDays。若用户已给出标题，title 可原样回填。",
            maxOutputTokens = 900,
            maxTokens = 900,
        )

        private val RECOMMENDATION_SPEC = LlmClient.SchemaSpec(
            schemaName = "capsule_recommendations",
            schema = buildJsonObject {
                put("type", "object")
                put("additionalProperties", false)
                putJsonArray("required") { add("items") }
                putJsonObject("properties") {
                    putJsonObject("items") {
                        put("type", "array")
                        put("minItems", 3)
                        put("maxItems", 8)
                        putJsonObject("items") {
                            put("type", "object")
                            put("additionalProperties", false)
                            putJsonArray("required") { add("title"); add("hint"); add("openInDays") }
                            putJsonObject("properties") {
                                putJsonObject("title") { put("type", "string") }
                                putJsonObject("hint") { put("type", "string") }
                                putJsonObject("openInDays") {
                                    put("type", "integer"); put("minimum", 1); put("maximum", 3650)
                                }
                            }
                        }
                    }
                }
            },
            systemPrompt = "你只返回严格 JSON 对象，不要 Markdown、代码块或解释。" +
                "JSON 必须包含数组字段 items，每项含字符串字段 title、hint 和整数字段 openInDays。",
            maxOutputTokens = 900,
            maxTokens = 900,
        )
    }
}

File: backends/aspnet/src/Services/LlmClient.cs
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using HelloTimePro.Aspnet.Config;

namespace HelloTimePro.Aspnet.Services;

/// <summary>
/// 结构化 JSON 生成的 LLM 客户端。日志规范、网关重试、CF 1010 等坑见 docs/dev-notes.md §3。
/// 契约测试默认 LLM_ENABLED=false 时此处立即抛异常，由上层走本地兜底；HTTP 路径不在测试链上。
/// 单例：复用一个 HttpClient，超时用 per-request CancellationToken 控制。
/// </summary>
public sealed class LlmClient
{
    private readonly AppConfig _config;
    private readonly ILogger<LlmClient> _log;
    private readonly HttpClient _http;

    public LlmClient(AppConfig config, ILogger<LlmClient> log)
    {
        _config = config;
        _log = log;
        _http = new HttpClient { Timeout = Timeout.InfiniteTimeSpan };
    }

    public sealed class LlmClientException : Exception
    {
        public int Status { get; }
        public LlmClientException(string message, int status = 0, Exception? cause = null) : base(message, cause)
            => Status = status;
    }

    private sealed record SchemaSpec(
        string SchemaName, JsonObject Schema, string SystemPrompt, int MaxOutputTokens, int MaxTokens);

    public Task<JsonObject> GenerateCapsuleSuggestion(string prompt) =>
        GenerateStructuredJson(prompt, SuggestionSpec);

    public Task<JsonObject> GenerateCapsuleRecommendations(string prompt) =>
        GenerateStructuredJson(prompt, RecommendationSpec);

    private async Task<JsonObject> GenerateStructuredJson(string prompt, SchemaSpec spec)
    {
        var llm = _config.Llm;
        if (!llm.Enabled || string.IsNullOrWhiteSpace(llm.ApiKey))
            throw new LlmClientException("LLM is disabled or missing API key");

        return llm.ApiStyle switch
        {
            "responses" => await GenerateWithResponses(prompt, spec),
            "auto" => await GenerateAuto(prompt, spec),
            _ => await GenerateWithChat(prompt, spec, disableThinking: true),
        };
    }

    private async Task<JsonObject> GenerateAuto(string prompt, SchemaSpec spec)
    {
        try
        {
            return await GenerateWithResponses(prompt, spec);
        }
        catch (LlmClientException e)
        {
            _log.LogInformation("Responses API unavailable ({Msg}); falling back to chat completions", e.Message);
            return await GenerateWithChat(prompt, spec, disableThinking: true);
        }
    }

    private async Task<JsonObject> GenerateWithResponses(string prompt, SchemaSpec spec)
    {
        var payload = new JsonObject
        {
            ["model"] = _config.Llm.Model,
            ["input"] = prompt,
            ["max_output_tokens"] = spec.MaxOutputTokens,
            ["text"] = new JsonObject
            {
                ["format"] = new JsonObject
                {
                    ["type"] = "json_schema",
                    ["name"] = spec.SchemaName,
                    ["strict"] = true,
                    ["schema"] = spec.Schema.DeepClone(),
                },
            },
        };
        var body = await PostJson(ResponsesUrl(), payload);
        return ParseJsonObject(ExtractResponsesText(body));
    }

    private async Task<JsonObject> GenerateWithChat(string prompt, SchemaSpec spec, bool disableThinking)
    {
        JsonObject BuildPayload(bool withThinking) => new()
        {
            ["model"] = _config.Llm.Model,
            ["messages"] = new JsonArray
            {
                new JsonObject { ["role"] = "system", ["content"] = spec.SystemPrompt },
                new JsonObject { ["role"] = "user", ["content"] = prompt },
            },
            ["max_tokens"] = spec.MaxTokens,
            ["thinking"] = withThinking ? new JsonObject { ["type"] = "disabled" } : null,
        };

        try
        {
            var body = await PostJson(ChatUrl(), BuildPayload(disableThinking));
            return ParseJsonObject(ExtractChatText(body));
        }
        catch (LlmClientException e) when (e.Status == 400)
        {
            // 某些网关不认 thinking 字段，去掉重试一次。
            var body = await PostJson(ChatUrl(), BuildPayload(withThinking: false));
            return ParseJsonObject(ExtractChatText(body));
        }
    }

    /// <summary>向 url POST JSON；瞬时网络/TLS 错误按配置重试，HTTP 4xx/5xx 与坏 JSON 不重试。</summary>
    private async Task<JsonObject> PostJson(string url, JsonObject payload)
    {
        var llm = _config.Llm;
        var bodyText = payload.ToJsonString();
        var attempts = Math.Max(1, llm.MaxRetries + 1);
        Exception? lastError = null;

        for (var attempt = 1; attempt <= attempts; attempt++)
        {
            _log.LogInformation("LLM request  model={Model} url={Url} attempt={Attempt}/{Attempts}",
                llm.Model, url, attempt, attempts);
            var start = Environment.TickCount64;
            using var cts = new CancellationTokenSource(TimeSpan.FromMilliseconds(llm.TimeoutMs));
            try
            {
                using var request = new HttpRequestMessage(HttpMethod.Post, url);
                request.Content = new StringContent(bodyText, Encoding.UTF8, "application/json");
                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", llm.ApiKey);
                request.Headers.Accept.ParseAdd("application/json");
                request.Headers.UserAgent.ParseAdd(llm.UserAgent);

                using var response = await _http.SendAsync(request, cts.Token);
                var elapsed = Environment.TickCount64 - start;
                var content = await response.Content.ReadAsStringAsync();
                var status = (int)response.StatusCode;
                if (status is < 200 or >= 300)
                {
                    _log.LogWarning("LLM error    model={Model} elapsed_ms={Elapsed} status={Status}",
                        llm.Model, elapsed, status);
                    throw new LlmClientException($"HTTP {status}: {content}", status);
                }

                JsonObject parsed;
                try
                {
                    parsed = JsonNode.Parse(content)!.AsObject();
                }
                catch (Exception e)
                {
                    _log.LogWarning("LLM error    model={Model} elapsed_ms={Elapsed} error=invalid-json", llm.Model, elapsed);
                    throw new LlmClientException("LLM response was not valid JSON", cause: e);
                }
                _log.LogInformation("LLM response model={Model} elapsed_ms={Elapsed} tokens={Tokens}",
                    llm.Model, elapsed, ExtractTokens(parsed));
                return parsed;
            }
            catch (LlmClientException)
            {
                throw; // HTTP / JSON 错误不重试。
            }
            catch (Exception e) when (e is HttpRequestException or TaskCanceledException or IOException)
            {
                var elapsed = Environment.TickCount64 - start;
                var willRetry = attempt < attempts;
                _log.LogWarning("LLM error    model={Model} elapsed_ms={Elapsed} error={Error}{Retry}",
                    llm.Model, elapsed, e.Message, willRetry ? " (will retry)" : "");
                lastError = e;
                if (willRetry) await Task.Delay(TimeSpan.FromMilliseconds(llm.RetryBackoffMs * attempt));
            }
        }
        throw new LlmClientException(lastError?.Message ?? "LLM request failed", cause: lastError);
    }

    private static string ExtractTokens(JsonObject body)
    {
        if (body["usage"] is not JsonObject usage) return "n/a";
        if (TryLong(usage["total_tokens"], out var total) && total > 0) return total.ToString();
        var sum = (TryLong(usage["input_tokens"], out var inp) ? inp : 0) +
                  (TryLong(usage["output_tokens"], out var outp) ? outp : 0);
        return sum > 0 ? sum.ToString() : "n/a";
    }

    private static bool TryLong(JsonNode? node, out long value)
    {
        value = 0;
        if (node is null) return false;
        try { value = node.GetValue<long>(); return true; }
        catch { return long.TryParse(node.ToString(), out value); }
    }

    private static string ExtractChatText(JsonObject body)
    {
        if (body["choices"] is not JsonArray choices || choices.Count == 0)
            throw new LlmClientException("LLM chat response missing choices");
        var content = choices[0]?["message"]?["content"]?.GetValue<string>();
        if (string.IsNullOrWhiteSpace(content))
            throw new LlmClientException("LLM chat response missing content");
        return content;
    }

    private static string ExtractResponsesText(JsonObject body)
    {
        if (body["output_text"] is JsonValue ot && ot.TryGetValue<string>(out var t) && !string.IsNullOrWhiteSpace(t))
            return t;
        if (body["output"] is JsonArray output)
        {
            foreach (var item in output)
            {
                if (item?["content"] is not JsonArray contentArr) continue;
                foreach (var entry in contentArr)
                {
                    if (entry?["text"] is JsonValue tv && tv.TryGetValue<string>(out var txt) && !string.IsNullOrWhiteSpace(txt))
                        return txt;
                }
            }
        }
        throw new LlmClientException("LLM response did not contain output text");
    }

    private static JsonObject ParseJsonObject(string raw)
    {
        var text = raw.Trim();
        if (text.StartsWith("```"))
        {
            text = System.Text.RegularExpressions.Regex.Replace(text, @"^```[a-zA-Z]*\s*", "");
            text = System.Text.RegularExpressions.Regex.Replace(text, @"\s*```$", "").Trim();
        }
        try
        {
            return JsonNode.Parse(text)!.AsObject();
        }
        catch
        {
            var start = text.IndexOf('{');
            var end = text.LastIndexOf('}');
            if (start < 0 || end <= start) throw new LlmClientException("LLM output was not valid JSON");
            return JsonNode.Parse(text[start..(end + 1)])!.AsObject();
        }
    }

    private string ResponsesUrl() => StripTrailingSlash(_config.Llm.BaseUrl) + "/responses";
    private string ChatUrl() => StripTrailingSlash(_config.Llm.BaseUrl) + "/chat/completions";

    private static string StripTrailingSlash(string value) =>
        string.IsNullOrWhiteSpace(value) ? "https://api.openai.com/v1" : value.TrimEnd('/');

    private static readonly SchemaSpec SuggestionSpec = new(
        SchemaName: "capsule_suggestion",
        Schema: new JsonObject
        {
            ["type"] = "object",
            ["additionalProperties"] = false,
            ["required"] = new JsonArray { "title", "content", "openInDays" },
            ["properties"] = new JsonObject
            {
                ["title"] = new JsonObject { ["type"] = "string" },
                ["content"] = new JsonObject { ["type"] = "string" },
                ["openInDays"] = new JsonObject { ["type"] = "integer", ["minimum"] = 1, ["maximum"] = 3650 },
            },
        },
        SystemPrompt: "你只返回严格 JSON 对象，不要 Markdown、代码块或解释。" +
                      "JSON 必须包含字符串字段 title、content 和整数字段 openInDays。若用户已给出标题，title 可原样回填。",
        MaxOutputTokens: 900,
        MaxTokens: 900);

    private static readonly SchemaSpec RecommendationSpec = new(
        SchemaName: "capsule_recommendations",
        Schema: new JsonObject
        {
            ["type"] = "object",
            ["additionalProperties"] = false,
            ["required"] = new JsonArray { "items" },
            ["properties"] = new JsonObject
            {
                ["items"] = new JsonObject
                {
                    ["type"] = "array",
                    ["minItems"] = 3,
                    ["maxItems"] = 8,
                    ["items"] = new JsonObject
                    {
                        ["type"] = "object",
                        ["additionalProperties"] = false,
                        ["required"] = new JsonArray { "title", "hint", "openInDays" },
                        ["properties"] = new JsonObject
                        {
                            ["title"] = new JsonObject { ["type"] = "string" },
                            ["hint"] = new JsonObject { ["type"] = "string" },
                            ["openInDays"] = new JsonObject { ["type"] = "integer", ["minimum"] = 1, ["maximum"] = 3650 },
                        },
                    },
                },
            },
        },
        SystemPrompt: "你只返回严格 JSON 对象，不要 Markdown、代码块或解释。" +
                      "JSON 必须包含数组字段 items，每项含字符串字段 title、hint 和整数字段 openInDays。",
        MaxOutputTokens: 900,
        MaxTokens: 900);
}
